inc/types/ManagedBuffer.h

```diff
@@ -0,0 +1,256 @@
+
+    /**
+     * Array access operation (modify).
+     *
+     * Called when a ManagedBuffer is dereferenced with a [] operation.
+     * Transparently map this through to the underlying payload for elegance of programming.
+     *
+     * Example:
+     * @code
+     * ManagedBuffer p1(16);
+     * p1[0] = 42;
+     * @endcode
+     */
+    uint8_t& operator [] (int i)
+    {
+        return ptr->payload[i];
+    }
+
+    /**
+      * Equality operation.
+      *
+      * Called when one ManagedBuffer is tested to be equal to another using the '==' operator.
+      *
+      * @param p The ManagedBuffer to test ourselves against.
+      * @return true if this ManagedBuffer is identical to the one supplied, false otherwise.
+      *
+      * Example:
+      * @code
+      *
+      * uint8_t buf = {13,5,2};
+      * ManagedBuffer p1(16);
+      * ManagedBuffer p2(buf, 3);
+      *
+      * if(p1 == p2)                    // will be true
+      *     uBit.display.scroll("same!");
+      * @endcode
+      */
+    bool operator== (const ManagedBuffer& p);
+
+    /**
+      * Sets the byte at the given index to value provided.
+      * @param position The index of the byte to change.
+      * @param value The new value of the byte (0-255).
+      * @return MICROBIT_OK, or MICROBIT_INVALID_PARAMETER.
+      *
+      * Example:
+      * @code
+      * ManagedBuffer p1(16);
+      * p1.setByte(0,255);              // Sets the first byte in the buffer to the value 255.
+      * @endcode
+      */
+    int setByte(int position, uint8_t value);
+
+    /**
+      * Determines the value of the given byte in the buffer.
+      *
+      * @param position The index of the byte to read.
+      * @return The value of the byte at the given position, or MICROBIT_INVALID_PARAMETER.
+      *
+      * Example:
+      * @code
+      * ManagedBuffer p1(16);
+      * p1.setByte(0,255);              // Sets the first byte in the buffer to the value 255.
+      * p1.getByte(0);                  // Returns 255.
+      * @endcode
+      */
+    int getByte(int position);
+
+    /**
+      * Gets number of bytes in this buffer
+      * @return The size of the buffer in bytes.
+      *
+      * Example:
+      * @code
+      * ManagedBuffer p1(16);
+      * p1.length();                 // Returns 16.
+      * @endcode
+      */
+    int length() const { return ptr->length; }
+
+    int fill(uint8_t value, int offset = 0, int length = -1);
+
+    ManagedBuffer slice(int offset = 0, int length = -1) const;
+
+    void shift(int offset, int start = 0, int length = -1);
+
+    void rotate(int offset, int start = 0, int length = -1);
+
+    int readBytes(uint8_t *dst, int offset, int length, bool swapBytes = false) const;
+
+    int writeBytes(int dstOffset, uint8_t *src, int length, bool swapBytes = false);
+
+    int writeBuffer(int dstOffset, const ManagedBuffer &src, int srcOffset = 0, int length = -1);
+
+    bool isReadOnly() const { return ptr->isReadOnly(); }
+};
+
+#endif
```

Comment:
coool... we should look to see if we can lay down some other managed types (e.g. PacketBuffer)  by having this more general type.